### ISSUE TITLE
proposal: install nxg to libdir

### ISSUE THREAD

**Issue by 0verse** [mini-rose/coffee#30]
If we do install nxg executable into the libdir  and then symlink it to the bindir, ie /usr/lib/coffee/nxg -> /usr/bin/nxg, we might not need a hardcoded lib path for nxg to have. nxg will use ./lib path instead of /usr/lib/coffee/lib.

Edit:
This is how it can be done with a meson example:
> install_symlink('target', pointing_to: '../bin/target', install_dir: '/usr/sbin')

In this case:
> nxg_file = join_paths(libdir, 'nxg')
> install_symlink('nxg', pointing_to: nxg_file, install_dir: bindir)

**Comment by bellrise** [comment]
In 268b94b, I changed the lib path to the "sysroot" path with /usr/lib/coffee being the default root path. This directory will be used not only for the standard libraries, but packages too in the future in the `pkg` directory (think of global npm packages).

I like the idea of linking the binary from, let's say, `/usr/lib/coffee/bin/nxg` to `/usr/bin/nxg`, it could allow for more precise version control (the binary could be actually named `nxg-0.6` and be symlinked to `nxg`. As for the hardcoded path, there is no problems with resolving anything from it with the meson --prefix option, so I don't think a change is needed.

**Comment by 0verse** [comment]
The latest commit resolved my issue, so thanks again for a great work. I agree on your view concerning the symlink to current nxg version.

**Comment by 0verse** [comment]
I've tested building, installing and running on Alpine Linux as well and so far so good. Just wanted to tell that "<sys/stat.h>" an Alpine and on my custom musl based distro includes "<bits/stat.h>" which causes error[1] that I have no idea why. Removing "<sys/stat.h>" from nxg/cc/build.c fixes the issue.

1:
> In file included from /usr/include/sys/stat.h:23,
>                  from ../src/cc/build.c:15:
> /usr/include/bits/stat.h:21:22: error: expected identifier or '(' before '[' token
>    21 |         long __unused[3];
>       |

**Comment by bellrise** [comment]
You seem to have a outdated bits/stat.h file, because my one from musl also has a __unused field and there are no syntax errors, and I can't remove sys/stat.h from build as it declares `mkdir()`.

**Comment by 0verse** [comment]
I see, but I have the latest musl version installed tough, v1.2.3. Maybe there is something else then.

**Comment by bellrise** [comment]
Can you show me your /usr/include/bits/stat.h file?

**Comment by 0verse** [comment]
```
/* copied from kernel definition, but with padding replaced
 * by the corresponding correctly-sized userspace types. */

struct stat {
    dev_t st_dev;
    ino_t st_ino;
    nlink_t st_nlink;

    mode_t st_mode;
    uid_t st_uid;
    gid_t st_gid;
    unsigned int    __pad0;
    dev_t st_rdev;
    off_t st_size;
    blksize_t st_blksize;
    blkcnt_t st_blocks;

    struct timespec st_atim;
    struct timespec st_mtim;
    struct timespec st_ctim;
    long __unused[3];
};
```

**Comment by bellrise** [comment]
How about sys/stat?

**Comment by 0verse** [comment]
#ifndef	_SYS_STAT_H
#define	_SYS_STAT_H
#ifdef __cplusplus
extern "C" {
#endif

#include <features.h>

#define __NEED_dev_t
#define __NEED_ino_t
#define __NEED_mode_t
#define __NEED_nlink_t
#define __NEED_uid_t
#define __NEED_gid_t
#define __NEED_off_t
#define __NEED_time_t
#define __NEED_blksize_t
#define __NEED_blkcnt_t
#define __NEED_struct_timespec

#include <bits/alltypes.h>

#include <bits/stat.h>

#define st_atime st_atim.tv_sec
#define st_mtime st_mtim.tv_sec
#define st_ctime st_ctim.tv_sec

#define S_IFMT  0170000

#define S_IFDIR 0040000
#define S_IFCHR 0020000
#define S_IFBLK 0060000
#define S_IFREG 0100000
#define S_IFIFO 0010000
#define S_IFLNK 0120000
#define S_IFSOCK 0140000

#define S_TYPEISMQ(buf)  0
#define S_TYPEISSEM(buf) 0
#define S_TYPEISSHM(buf) 0
#define S_TYPEISTMO(buf) 0

#define S_ISDIR(mode)  (((mode) & S_IFMT) == S_IFDIR)
#define S_ISCHR(mode)  (((mode) & S_IFMT) == S_IFCHR)
#define S_ISBLK(mode)  (((mode) & S_IFMT) == S_IFBLK)
#define S_ISREG(mode)  (((mode) & S_IFMT) == S_IFREG)
#define S_ISFIFO(mode) (((mode) & S_IFMT) == S_IFIFO)
#define S_ISLNK(mode)  (((mode) & S_IFMT) == S_IFLNK)
#define S_ISSOCK(mode) (((mode) & S_IFMT) == S_IFSOCK)

#ifndef S_IRUSR
#define S_ISUID 04000
#define S_ISGID 02000
#define S_ISVTX 01000
#define S_IRUSR 0400
#define S_IWUSR 0200
#define S_IXUSR 0100
#define S_IRWXU 0700
#define S_IRGRP 0040
#define S_IWGRP 0020
#define S_IXGRP 0010
#define S_IRWXG 0070
#define S_IROTH 0004
#define S_IWOTH 0002
#define S_IXOTH 0001
#define S_IRWXO 0007
#endif

#define UTIME_NOW  0x3fffffff
#define UTIME_OMIT 0x3ffffffe

int stat(const char *__restrict, struct stat *__restrict);
int fstat(int, struct stat *);
int lstat(const char *__restrict, struct stat *__restrict);
int fstatat(int, const char *__restrict, struct stat *__restrict, int);
int chmod(const char *, mode_t);
int fchmod(int, mode_t);
int fchmodat(int, const char *, mode_t, int);
mode_t umask(mode_t);
int mkdir(const char *, mode_t);
int mkfifo(const char *, mode_t);
int mkdirat(int, const char *, mode_t);
int mkfifoat(int, const char *, mode_t);

#if defined(_XOPEN_SOURCE) || defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
int mknod(const char *, mode_t, dev_t);
int mknodat(int, const char *, mode_t, dev_t);
#endif

int futimens(int, const struct timespec [2]);
int utimensat(int, const char *, const struct timespec [2], int);

#if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
int lchmod(const char *, mode_t);
#define S_IREAD S_IRUSR
#define S_IWRITE S_IWUSR
#define S_IEXEC S_IXUSR
#endif

#if defined(_LARGEFILE64_SOURCE) || defined(_GNU_SOURCE)
#define stat64 stat
#define fstat64 fstat
#define lstat64 lstat
#define fstatat64 fstatat
#define blkcnt64_t blkcnt_t
#define fsblkcnt64_t fsblkcnt_t
#define fsfilcnt64_t fsfilcnt_t
#define ino64_t ino_t
#define off64_t off_t
#endif

#if _REDIR_TIME64
__REDIR(stat, __stat_time64);
__REDIR(fstat, __fstat_time64);
__REDIR(lstat, __lstat_time64);
__REDIR(fstatat, __fstatat_time64);
__REDIR(futimens, __futimens_time64);
__REDIR(utimensat, __utimensat_time64);
#endif

#ifdef __cplusplus
}
#endif
#endif

**Comment by bellrise** [comment]
No idea what the problem is, it all looks good to me. I have the exact same musl version and files :/

**Comment by 0verse** [comment]
it could be my machine doing strange thing, since I've done a chroot setup on my machine when using alpine linux.

**Comment by 0verse** [comment]
@bellrise could it be because of this:
#define __unused __attribute__((unused)) (https://github.com/mini-rose/coffee/blob/master/nxg/include/nxg/nxg.h#L40)

**Comment by bellrise** [comment]
Ahh I see, I'll make a change so that the attribute is called differently.

**Comment by bellrise** [comment]
Fixed as of 2bb4793

**Comment by 0verse** [comment]
well done!
